lib/src/fty_shm.cc

```diff
@@ -336,191 +311,249 @@ int fty_shm_read_family(const char* family, std::string asset, std::string type,
+            value = static_cast<char*>(zhash_next(aux));
         }
     }
-    if (fclose(file) < 0)
-        return -1;
-
-    Publisher::publishMetric(metric); //mqtt-pub
-    return 0;
-}
 
-int fty::shm::write_metric(fty_proto_t* metric)
-{
-    char filename[PATH_MAX];
+    fty::shm::Publisher::publishMetric(metric); //mqtt-pub
 
-    if (prepare_filename(filename, fty_proto_name(metric), strlen(fty_proto_name(metric)), fty_proto_type(metric),
-            strlen(fty_proto_type(metric)), FTY_SHM_METRIC_TYPE) < 0)
+    if (fclose(file) < 0) {
         return -1;
-    return write_metric_data(filename, metric);
+    }
+    return 0;
 }
 
 int fty_shm_write_metric_proto(fty_proto_t* metric)
 {
-    char filename[PATH_MAX];
+    return write_metric_data(metric);
+}
 
-    if (prepare_filename(filename, fty_proto_name(metric), strlen(fty_proto_name(metric)), fty_proto_type(metric),
-            strlen(fty_proto_type(metric)), FTY_SHM_METRIC_TYPE) < 0)
-        return -1;
-    return write_metric_data(filename, metric);
+int fty::shm::write_metric(fty_proto_t* metric)
+{
+    return fty_shm_write_metric_proto(metric);
 }
 
-int fty::shm::write_metric(
-    const std::string& asset, const std::string& metric, const std::string& value, const std::string& unit, int ttl)
+int fty::shm::write_metric(const std::string& asset, const std::string& metric, const std::string& value, const std::string& unit, int ttl)
 {
-    char filename[PATH_MAX];
+    if (asset.empty() || metric.empty() || value.empty()) {
+        return -1;
+    }
+
+    if (ttl < 0) {
+        ttl = 0;
+    }
 
-    if (prepare_filename(
-            filename, asset.c_str(), asset.length(), metric.c_str(), metric.length(), FTY_SHM_METRIC_TYPE) < 0)
+    fty_proto_t *proto_metric = fty_proto_new(FTY_PROTO_METRIC);
+    if (!proto_metric) {
         return -1;
-    return write_value(filename, value.c_str(), unit.c_str(), ttl);
+    }
+
+    fty_proto_set_name(proto_metric, "%s", asset.c_str());
+    fty_proto_set_type(proto_metric, "%s", metric.c_str());
+    fty_proto_set_value(proto_metric, "%s", value.c_str());
+    fty_proto_set_unit(proto_metric, "%s", unit.c_str());
+    fty_proto_set_ttl(proto_metric, static_cast<uint32_t>(ttl));
+
+    int ret = fty::shm::write_metric(proto_metric);
+
+    fty_proto_destroy(&proto_metric);
+
+    return ret;
 }
 
 int fty::shm::read_metric_value(const std::string& asset, const std::string& metric, std::string& value)
 {
-    char        filename[PATH_MAX];
-    std::string dummy;
+    char* value_s = NULL;
+    int ret = fty_shm_read_metric(asset.c_str(), metric.c_str(), &value_s, NULL);
+    if (ret == 0) { // ok
+        value = value_s ? value_s : "";
+    }
 
-    if (prepare_filename(
-            filename, asset.c_str(), asset.length(), metric.c_str(), metric.length(), FTY_SHM_METRIC_TYPE) < 0)
-        return -1;
-    return read_value(filename, value, dummy, false);
+    // cleanup
+    if (value_s) {
+        free(value_s);
```

Comment:
delete ?

Reply:
didn't understand what you mean

Reply:
Sorry it is ok, never see what you use strdup for allocation (malloc)